### PR TITLE
java-cdk-destination: implement logic to update destination stats in state message from async framework

### PR DESCRIPTION
**Pull request by subodh1810**
Issue : https://github.com/airbytehq/airbyte/issues/33863

Tested this locally using Postgres source and snowflake destination. The stats are being updated and returned as part of the state! 

Sample state messages 

```
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models_random"},"stream_state":{"version":2,"state_type":"ctid","ctid":"(0,2)","incremental_state":{"state_type":"cursor_based","version":2,"stream_name":"models_random","stream_namespace":"models_schema_random","cursor_field":["id_random"],"cursor":"16000","cursor_record_count":1},"relation_filenode":16394}},"destinationStats":{"recordCount":2.0}}}
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models"},"stream_state":{"version":2,"state_type":"ctid","ctid":"(0,2)","incremental_state":{"state_type":"cursor_based","version":2,"stream_name":"models","stream_namespace":"models_schema","cursor_field":["id"],"cursor":"16","cursor_record_count":1},"relation_filenode":16388}},"destinationStats":{"recordCount":2.0}}}
2024-01-08 12:34:43 platform > starting state flush thread for connectionId 4657ad5c-3e7a-44f6-922c-e766c718b956
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models"},"stream_state":{"version":2,"state_type":"ctid","ctid":"(0,4)","incremental_state":{"state_type":"cursor_based","version":2,"stream_name":"models","stream_namespace":"models_schema","cursor_field":["id"],"cursor":"16","cursor_record_count":1},"relation_filenode":16388}},"destinationStats":{"recordCount":2.0}}}
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models_random"},"stream_state":{"version":2,"state_type":"ctid","ctid":"(0,4)","incremental_state":{"state_type":"cursor_based","version":2,"stream_name":"models_random","stream_namespace":"models_schema_random","cursor_field":["id_random"],"cursor":"16000","cursor_record_count":1},"relation_filenode":16394}},"destinationStats":{"recordCount":2.0}}}
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models"},"stream_state":{"version":2,"state_type":"cursor_based","stream_name":"models","stream_namespace":"models_schema","cursor_field":["id"],"cursor":"16","cursor_record_count":1}},"destinationStats":{"recordCount":2.0}}}
2024-01-08 12:34:43 destination > 2024-01-08 12:34:43 INFO i.a.c.i.d.FlushWorkers(emitStateMessages):244 - STATE_MESSAGE : {"type":"STATE","state":{"type":"STREAM","stream":{"stream_descriptor":{"name":"models_random"},"stream_state":{"version":2,"state_type":"cursor_based","stream_name":"models_random","stream_namespace":"models_schema_random","cursor_field":["id_random"],"cursor":"16000","cursor_record_count":1}},"destinationStats":{"recordCount":2.0}}}
```

### CODE REVIEW

File: airbyte-integrations/connectors/destination-snowflake/build.gradle
Comment (subodh1810) [inline airbyte-integrations/connectors/destination-snowflake/build.gradle:9]: temporary commit to test the changes with snowflake source, will revert this before merging

**Review by jdpgrailsdev — APPROVED** [review]
:shipit: LGTM, as long as the `useLocalCdk` change is reverted prior to merging per your comment!

File: airbyte-integrations/connectors/destination-snowflake/build.gradle
Comment (evantahler) [reply airbyte-integrations/connectors/destination-snowflake/build.gradle:9]: Don't forget to bump the CDK version and publish in this PR as well!

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/PartialStateWithDestinationStats.java
Comment (evantahler) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/PartialStateWithDestinationStats.java:8]: I would have expected to see a version bump of the protocol package to get this new `AirbyteStateStats` object.

Comment (subodh1810) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/PartialStateWithDestinationStats.java:8]: Well turns out you dont! haha! The versioning thing is different. v0 is still in prod. We need to put in work to migrate everything to v1.

File: airbyte-integrations/connectors/destination-snowflake/build.gradle
Comment (subodh1810) [reply airbyte-integrations/connectors/destination-snowflake/build.gradle:9]: Yup! I plan to publish the cdk in this PR and will raise a separate PR to bump the version for snowflake destination

**Comment by subodh1810** [comment]
Snowflake tests are passing with these changes

Screenshot 2024-01-08 at 10 35 46 PM

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:189]: why is this a double...? shouldn't it always be an integer

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:219]: is `increment` ever negative?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java:163]: do we have a test that verifies we still checkpoint correctly on streams with null namespace after this change? AsyncStreamConsumerTest seems (at a glance) to only have tests with nonnull namespace

(my guess is that checkpointing is broken after this, but I have no evidence :shrug:)

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (benmoriceau) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:219]: I don't think this is possible since it is called with a 0 value. Unfortunatelly, there is no ulong in java

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java
Comment (benmoriceau) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java:163]: Why do we need to override the namespace here?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (subodh1810) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:189]: So as discussed over call this is a double because the platform changes require it to be long

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java
Comment (subodh1810) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/state/GlobalAsyncStateManager.java:219]: No it wont in the current flow. This is just a safety check that i put in place

File: airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java
Comment (subodh1810) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/java/io/airbyte/cdk/integrations/destination_async/AsyncStreamConsumer.java]: So I made the change to remove this mutation. Take a look at my comment in class `GlobalAsyncStateManager`
```
/**
   * If the user has selected the Destination Namespace as the Destination default while setting up
   * the connector, the platform sets the namespace as null in the StreamDescriptor in the
   * AirbyteMessages (both record and state messages). The destination checks that if the namespace is
   * empty or null, if yes then re-populates it with the defaultNamespace. See
   * {@link io.airbyte.cdk.integrations.destination_async.AsyncStreamConsumer#accept(String,Integer)}
   * But destination only does this for the record messages. So when state messages arrive without a
   * namespace and since the destination doesn't repopulate it with the default namespace, there is a
   * mismatch between the StreamDescriptor from record messages and state messages. That breaks the
   * logic of the state management class as {@link descToStateIdQ} needs to have consistent
   * StreamDescriptor. This is why while trying to extract the StreamDescriptor from state messages,
   * we check if the namespace is null, if yes then replace it with defaultNamespace to keep it
   * consistent with the record messages.
   */
```

File: airbyte-cdk/java/airbyte-cdk/core/src/test/java/io/airbyte/cdk/integrations/destination_async/buffers/BufferEnqueueTest.java
Comment (edgao) [inline airbyte-cdk/java/airbyte-cdk/core/src/test/java/io/airbyte/cdk/integrations/destination_async/buffers/BufferEnqueueTest.java:22]:
```suggestion
  private static final String DEFAULT_NAMESPACE = "foo_namespace";
```

(and also for the other test(s))

**Review by edgao — APPROVED** [review]
one nit on the tests to make them match realistic behavior.

I also submitted https://github.com/airbytehq/airbyte/issues/34051 to deal with the record message mutations better.

otherwise lgtm!

**Comment by edgao** [comment]
I think this broke DATs, but was undetected because Snowflake disabled legacy DATs. E.g. postgres is now falling tests https://github.com/airbytehq/airbyte/pull/34071

**Comment by gisripa** [comment]
@edgao  Mine started failing as soon as i started doing localCdk https://github.com/airbytehq/airbyte/pull/33875 How are the changes related to DATs ?

**Comment by edgao** [comment]
I think
1. this change updates async destination connectors to emit state messages with DestinationStats
2. if you merged latest master + enabled local CDK, you'll build the connector with this change
3. DATs are doing an assertEquals, where the expected state message does not have DestinationStats. But now your connector has non-null DestinationStats, so the assert fails

**Comment by subodh1810** [comment]
@edgao oops! Can we tackle it as we release individual connectors? I really don't want to fix every destination as part of this work!

**Comment by edgao** [comment]
my guess is it's a one-time fix for all connectors - there should be some way to update the assert to basically `assertEquals(expectedStateMessage, actualStateMessage.withDestinationStats(null))`

if there are postgres-specific failures then sure we can fix those individually

**Pull request merged**
